### ISSUE TITLE
Resource Calculation needs to lock and queue

### ISSUE THREAD

**Issue by Phoscur** [ugml/api#82]
As mentioned in Discord:
> I wonder if we can reuse some of my resource calculation code for your api @mamen , thats missing for a v1, we might keep kafka integration out for v2 to do on demand calculations first?
> currently there is no implementation to update resources on a planet to a given time, and the issue is more complex that you might think (see my blog^^)
> Resource calculation that considers fuel and tank limits:
> https://github.com/Phoscur/Phlame/tree/master/src/model/resources

The current implementation here goes without queueing and simply locks to a single job
https://github.com/ugml/api/blob/8a6cc945920c6505b20831246cf29d21a9aaaaca/src/routes/BuildingsRouter.ts#L168

If you consider that you have to calculate other multiple time intervals anyways (replaying multiple events, e.g. resource calculation before and after a building finished), it makes sense to solve this completely right away.

It doesn't mean that events have to be handled at their corresponding realtime! It is possible to calculate lazily on demand instead.

This is also a part to be (re)used server- and clientside to validate if the unit is to be ready to be built.

**Comment by mamen** [comment]
You are right, the whole build-system currently is more or less a 1:1 implementation of the old php-clients from over 10 years ago. The system needs to be adapted to be more modern and flexible. This also is a bigger part in the whole event-handling process.

We should start with a concept of how we want the build-system should look like and then start from there. What do you think about that?

**Comment by Phoscur** [comment]
Well the build queue (no matter how many items are in it), is always dependent on resource recalculation - different queues run in parallel:
1. buildings (building finish usually impacts resource calculation)
2. tech (no queuing, rarely impacts resource calculation)
3. hangar (parallel queueing, not changing resource production (yet))
but realistic queuing would allow for:
- error control when resources get stolen
- early&lazy queuing foreseeing resources or waiting for them
- resource consuming / feeding build processes 
- pausing progress (e.g. in case of power failure)

**Comment by mamen** [comment]
> 3. hangar (parallel queueing, not changing resource production (yet))

This is not 100% correct. Solar Satellites can have impacts on resource calculation if the planet had a negative energy-balance before the satellites were built.

This all boils down to one core component, the event-handler. In general, we need a system that can update parts of the game-state in (almost) real time or later, when not needed immediately. But for all of this, we need a good message-broker system to manage the amount of data to be calculated and a good architecture for the event-handler. This will be the next big component of this project, after the API is "finished". (Finished in exclamation-marks because no project is ever finished 😁 )

**Comment by Phoscur** [comment]
> This is not 100% correct.
You are totally right, I edited my comment, so all kinds of production need to be considered for resource calculation. 

Still, this is read-only for anyone but the owner and the battle system. It should be part of the prediction engine in the client too, so don't focus too much on server technology here, which ultimately only validates, delegates and persists.

My approach would be to try to improve on my old resource models, trying to stuff as much of business-logic in there, so the server is just a thin wrapper, a collection of route controllers which form some kind of event handler, yes. 

We could also focus on values rather than models, that might produce faster results, but I am a fan of the domain-driven approach. The event handler on the server-side, in the end, is just the outer while loop around the resources and buildings/tech/hanger `while (time < Date.now()) {`.
But the client also wants to calculate and check future events (can I queue this building safely?).

Btw do we want a wait option instead of canceling a building queue, or are resources consumed when queuing like with ships?
